providers/core.tcpsocket.js

```diff
@@ -26,35 +26,44 @@ Socket_firefox.prototype.getInfo = function(continuation) {
 Socket_firefox.prototype.close = function(continuation) {
   if (!this.clientSocket && !this.serverSocket) {
     continuation(undefined, {
-    "errcode": "SOCKET_CLOSED",
-    "message": "Cannot close non-connected socket"
+      "errcode": "SOCKET_CLOSED",
+      "message": "Cannot close non-connected socket"
     });
   } else if(this.clientSocket) {
-    this.clientSocket.close();
+    this.clientSocket.close(continuation);
   } else if (this.serverSocket) {
-    this.serverSocket.disconnect();
+    this.serverSocket.disconnect(continuation);
   } else {
     continuation(undefined, {
       'errcode': 'SOCKET_CLOSED',
       'message': 'Socket Already Closed, or was never opened'
     });
-    return;
   }
-  continuation();
 };
 
 // TODO: handle failures.
 Socket_firefox.prototype.connect = function(hostname, port, continuation) {
   this.clientSocket = new ClientSocket();
-  this.clientSocket.onDisconnect = function(err) {
-    this.dispatchEvent("onDisconnect", err);
-  }.bind(this);
+  this.clientSocket.onDisconnect = this.dispatchDisconnect.bind(this); 
   this.clientSocket.setOnDataListener(this._onData.bind(this));
   this.clientSocket.connect(hostname, port, false, continuation);
   this.hostname = hostname;
   this.port = port;
 };
 
+Socket_firefox.prototype.dispatchDisconnect = function(continuation, err) {
+  if (typeof err === 'undefined') {
+    err = {
+      'errcode': 'CONNECTION_CLOSED',
+      'message': 'Connection closed gracefully'
+    };
+  }
+  this.dispatchEvent('onDisconnect', err);
+  if (typeof continuation === 'function') {
```

Comment:
this appears to never be called without a continuation, so you should be always able to call continuation(), and don't need to verify it's a function.